### PR TITLE
[Merged by Bors] - Implement resource requests and limits for Trino pods

### PR DESCRIPTION
**Pull request by maltesander**
# Description

- PVC, memory and CPU requests and limits now configurable
- added new integration tests folder "resources"

closes https://github.com/stackabletech/trino-operator/issues/248

## Review Checklist

- [x] Code contains useful comments
- [x] CRD change approved (or not applicable)
- [x] (Integration-)Test cases added (or not applicable)
- [x] Documentation added (or not applicable)
- [x] Changelog updated (or not applicable)
- [x] Cargo.toml only contains references to git tags (not specific commits or branches)
- [x] Helm chart can be installed and deployed operator works (or not applicable)

Once the review is done, comment `bors r+` (or `bors merge`) to merge. [Further information](https://bors.tech/documentation/getting-started/#reviewing-pull-requests)

### CODE REVIEW

**Review by sbernauer — COMMENTED** [review]
Only some minor comments

**Comment by sbernauer** [comment]
In https://ci.stackable.tech/view/02%20Operator%20Tests%20(custom)/job/trino-operator-it-custom/18/consoleFull we have test failure such as
`https://ci.stackable.tech/view/02%20Operator%20Tests%20(custom)/job/trino-operator-it-custom/18/consoleFull`

File: rust/crd/src/lib.rs [hunk 1]
```diff
@@ -263,14 +269,45 @@ impl FromStr for TrinoRole {
...
+                min: Some(Quantity("100m".to_owned())),
+                max: Some(Quantity("1".to_owned())),
+            },
+            memory: MemoryLimits {
+                limit: Some(Quantity("2Gi".to_owned())),
+                runtime_limits: NoRuntimeLimits {},
+            },
+            storage: TrinoStorageConfig {
+                data: PvcConfig {
+                    capacity: Some(Quantity("2Gi".to_owned())),
```
Comment (sbernauer) [inline rust/crd/src/lib.rs:304]: The docs say 1Gi. Somehow the test see to default to 1Gi as well...

File: docs/modules/ROOT/pages/usage.adoc [hunk 2]
```diff
@@ -488,3 +488,91 @@ workers:
...
+workers:
+  roleGroups:
+    default:
+      config:
+        resources:
+          memory:
+            limit: '2Gi'
+----
+
+In this example, each Trino container in the "default" group will have a maximum of 2 gigabytes of memory. To be more precise, these memory limits apply to the containers running the Trino daemons but not to any sidecar containers that are part of the pod.
```
[the remark this answers is not in the document]
Comment (maltesander) [reply docs/modules/ROOT/pages/usage.adoc:532]: But there is only one container running trino ?

File: rust/crd/src/lib.rs (hunk 1 above)
Comment (maltesander) [reply rust/crd/src/lib.rs:304]: I think the crd was not updated correctly or something and therefore ignored...will invstigate

**Comment by sbernauer** [comment]
Ahh you did make a small mistake when triggering your custom test: You have to pick the `0.5.0-pr270` version of the Trino operator, not just the current nightly (bear in mind to wait that the github action building your `0.5.0-pr270` image + helm-chart has finished before starting the tests)
Re-triggered it with your pr version: https://ci.stackable.tech/view/02%20Operator%20Tests%20(custom)/job/trino-operator-it-custom/19/console

File: docs/modules/ROOT/pages/usage.adoc [hunk 3]
```diff
@@ -488,3 +488,91 @@ workers:
 ----
 
 Here too, overriding properties such as `http-server.https.port` will lead to broken installations.
+
+=== Storage for data volumes
+
+You can mount volumes where data is stored by specifying https://kubernetes.io/docs/concepts/storage/persistent-volumes[PersistentVolumeClaims] for each individual role or role group:
```
[the remark this answers is not in the document]
Comment (maltesander) [reply docs/modules/ROOT/pages/usage.adoc:494]: done

File: docs/modules/ROOT/pages/usage.adoc [hunk 4]
```diff
@@ -488,3 +488,91 @@ workers:
...
+      config:
+        resources:
+          storage:
+            data:
+              capacity: 3Gi
+----
+
+In the above example, all Trino workers in the default group will store data (the location of the property `--data-dir`) on a `3Gi` volume. Additional role groups not specifying any resources will inherit the config provided on the role level (`2Gi` volume). This works the same for memory or CPU requests.
+
+By default, in case nothing is configured in the custom resource for a certain role group, each Pod will have a `1Gi` large local volume mount for the data location.
```
[the remark this answers is not in the document]
Comment (maltesander) [reply docs/modules/ROOT/pages/usage.adoc:515]: done

File: rust/crd/src/lib.rs [hunk 5]
```diff
@@ -263,14 +269,45 @@ impl FromStr for TrinoRole {
 #[serde(rename_all = "camelCase")]
 pub struct TrinoClusterStatus {}
 
-#[derive(Clone, Debug, Default, Deserialize, Eq, JsonSchema, PartialEq, Serialize)]
+#[derive(Clone, Debug, Default, Deserialize, Merge, JsonSchema, PartialEq, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct TrinoStorageConfig {
+    #[serde(default)]
+    pub data: PvcConfig,
```
[the remark this answers is not in the document]
Comment (maltesander) [reply rust/crd/src/lib.rs:276]: I think data is fine. Its not an important mount like e.g. in nifi with flow etc. And logs / config is data afterall in my opinion.

File: docs/modules/ROOT/pages/usage.adoc (hunk 2 above)
[the remark this answers is not in the document]
Comment (sbernauer) [reply docs/modules/ROOT/pages/usage.adoc:532]: Ah, misread it. Let me update

**Review by sbernauer — APPROVED** [review]
LGTM when https://ci.stackable.tech/view/02%20Operator%20Tests%20(custom)/job/trino-operator-it-custom/19/console succeeds

**Comment by maltesander** [comment]
bors merge

**Pull request closed without merging**